### PR TITLE
Extensions API

### LINKED ISSUE #390: Extensions API
# Overview
We need to design and build an API layer for extensions, such that the functions are callable from within Deno.

# Acceptance Criteria
- [ ] All functions and types extensions may need are available and can be exposed to Deno.

### PR DESCRIPTION
**Pull request by andreaphylum**
Closes #390, blocked by #353.

The functions in `api.rs` are to be exposed by Deno. There is a mechanism for lazily evaluating common dependencies that are internal implementation details (currently: `Config` and `PhylumApi`).

### CODE REVIEW

**Review by cd-work — CHANGES_REQUESTED** [review]

File: .cargo/config
Comment (andreaphylum) [reply .cargo/config:3]: Well, it could either read as a noun (as in "[apply] lints") or as a verb "lint". No strong preference on my end.

Comment (cd-work) [reply .cargo/config:3]: Commands should generally be verbs. This is true for all `cargo` commands I know.

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs:30]: Instantiating things like `PhylumApi` is expensive as it touches network and not every extension may need to do that, but we are going to have to provide these values ahead of time (via `OpState`). Having those lazy means we don't pay the cost upfront.

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs:78]: It is an actionable value, meaning it could be an input to other calls (e.g. suppose the extension wants to implement a polling mechanism to check for status).

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs:91]: Same reasoning as above; I'm totally ok with a different naming convention, but in that case we need to document the correspondence between CLI endpoints and extensions API.

File: cli/src/commands/extensions/extension.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/extension.rs:13]: I went with the Rust default of capitalizing (`Cargo.toml`). No strong preference here either.

File: cli/src/commands/extensions/extension.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/extension.rs:16]: It's a safe default which makes things easy to check and enforces a consistent styling to a degree, solving at the same time conflict issues in contexts like case-insensitive filesystems (i.e. Windows).

File: cli/src/commands/extensions/extension.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/extension.rs:66]: This would be more pertinent to #353. I think checking specifically for XDG would add unnecessary complexity, but we can consider it if there are strong reasons why we may want to explicitly not set permissions.

File: cli/src/config.rs
Comment (andreaphylum) [reply cli/src/config.rs:210]: IIRC, the few places where those two functions are used, are using them in a slightly different way that warranted a bit of discrepancy. I can go back and check.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [reply cli/src/commands/extensions/api.rs:34]: If anything I'd just make it `pub` at that point. The concept of `pub(super)` is just unnecessarily complicated since the relative positioning requires unnecessary cognitive overhead to resolve. So either private, `pub` or `pub(crate)` should be used instead (imo). Overly specific visibility modifiers rarely provide much benefit.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [reply cli/src/commands/extensions/api.rs:30]: I feel like it's going to be much simpler to just always create the `PhylumApi`. Should there be any performance concerns, which would obviously also affect our own subcommands where we don't do this, we can optimize in the future.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [reply cli/src/commands/extensions/api.rs:78]: Is there a way to query for the status with this API though? Otherwise the use of this function just seems somewhat limited. The ID by itself is useless unless you have something you can do with it.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [reply cli/src/commands/extensions/api.rs:91]: There's a big difference between `phylum auth token --bearer` and `phylum_auth_token_bearer`. I don't think it's appropriate to mimic the literal words of the CLI, since CLIs are a syntax with certain meaning. Instead of translating that word-for-word, we should translate its meaning.

So for the example `phylum auth token --bearer`, we're clearly not trying to get `auth token bearer` we're telling the CLI, that we want the token of the type bearer.

File: cli/src/commands/extensions/extension.rs
Comment (cd-work) [reply cli/src/commands/extensions/extension.rs:16]: Did you take inspiration from other tools like GitHub's CLI, or did you just pick the minimal possible set of characters you deemed necessary?

File: cli/src/config.rs
Comment (cd-work) [reply cli/src/config.rs:210]: They don't seem to behave significantly differently looking at the code though. Other than error handling they are the same.

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs:34]: I agree that `pub(super)` may be not totally clear and require cognitive overhead, but more general `pub` exposures also go beyond the intended scope of those functions which is very narrow by intentional design (i.e. the Deno runtime should be the only client). Would having those functions be `pub(crate)` and the module being private under `extensions` be a good solution?

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs:30]: We don't initialize `PhylumApi` for each of the CLI's commands. On my VM there is a latency of a couple seconds for commands that do use the API compared to the ones that don't, which is high enough in my opinion to warrant opting into it only when necessary (granted, it might depend on my specific networking configuration, but we should not assume the happy path for all the users).

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs:78]: We should probably add that. As I mentioned on Notion, the API surface is a little limited as it is, only exposing equivalent CLI endpoints. I suggest that once the general scaffolding for running an extension which invokes API functions is in place, we start expanding the API surface, but we should have a clearer idea about how to pass data around first.

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs:81]: Good point -- I will make that change. So, let's not merge this in just yet, let's use this branch as a base for further work.

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs:91]: That makes sense, but we still do want to retain correspondence with CLI endpoints. I will add doc comments that say "This is the equivalent of `phylum foo --bar`".

File: cli/src/commands/extensions/extension.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/extension.rs:16]: The latter -- I think this does not impose an uncomfortable restriction, as e.g. most package managers follow that pattern, on the users and is at the same time easy to check for. If we can come up with compelling arguments for doing otherwise (being less or even _more_ restrictive, like say [npm](https://github.com/npm/validate-npm-package-name#naming-rules)), I'm not opposed to changing this.

File: docs/phylum_extension_add.md
Comment (andreaphylum) [reply docs/phylum_extension_add.md:9]: We don't have a final answer to that yet, but it's a good point. We may want to capture this in an issue.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [reply cli/src/commands/extensions/api.rs]: Yeah that seems fine to me.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [reply cli/src/commands/extensions/api.rs:30]: Yeah fair enough, there certainly is some delay at the very least. And another point is the authentication that might be required when the user is not signed in already.

Though I believe that would probably look pretty wonky when done lazily? We wouldn't want to interrupt extension's CLI output with our CLI's auth messages.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [reply cli/src/commands/extensions/api.rs:78]: The CLI's analyze command right now does more than just return the Job ID though, right? So I think when trying to appropriately mimic its API surface, we should return the result of the analysis?

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs]: Agreed, documentation is probably the best way to go.

File: cli/src/commands/extensions/extension.rs
Comment (cd-work) [reply cli/src/commands/extensions/extension.rs:16]: Having the first character be a letter is likely a prerequisite to not have clap yell at us, so that's a great point.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [inline cli/src/commands/extensions/api.rs:53]: I'd much prefer value/inner over just "t".

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [inline cli/src/commands/extensions/api.rs:33]: I don't think it's necessary to state the same thing twice, especially because the struct already makes this rather obvious.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [inline cli/src/commands/extensions/api.rs:74]: Having this method on `ExtensionState` seems **very** strange to me. There's basically nothing tying the `ExtensionState` struct to `OpState`, in fact `OpState` might not even have an `ExtensionState` stored.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [inline cli/src/commands/extensions/api.rs:100]: This seems like a strange place for this comment. What is the goal of documenting this here and how is it supposed to help someone writing new stuff?

I'd imagine most people would just copy/paste some random small excerpt from another extension API anyway, so I don't see why the steps need to be outlined in a comment here.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [inline cli/src/commands/extensions/api.rs:102]: What's the point of pinning the state when `ExtensionState::borrow_from` doesn't require state to be pinned?

**Review by cd-work — CHANGES_REQUESTED** [review]

File: cli/tests/fixtures/extensions/api-extension/phylum-api.ts
Comment (andreaphylum) [reply cli/tests/fixtures/extensions/api-extension/phylum-api.ts:4]: I tried to do that in a few potential variations, but in the end a valid URL was required every time -- iiuc Deno has no concept that would allow `import 'phylum-api'` directly, it has to be an existing and resolvable URL. The [FAQ](https://deno.land/manual/linking_to_external_code#faq) seems to also corroborate this, while not stating it explicitly. I tried specifying e.g. `./phylum-api.ts` and then adding a code module to the runtime via `include_str!` but Deno errored mentioning that the file did not exist. I guess we can dig deeper on this, e.g. how is the top-level scope filled (the one that contains `Deno`)? Maybe we could check if we can evaluate some JS code that fills that scope?

File: cli/tests/extensions.rs
Comment (andreaphylum) [reply cli/tests/extensions.rs:290]: Yeah, indeed -- the test is poorly set up since at this point I'd only be checking that nothing blows up, really. We need a mocking setup for these tests anyway, which I think we should address as part of a dedicated ticket.

File: cli/tests/fixtures/extensions/api-extension/PhylumExt.toml
Comment (andreaphylum) [reply cli/tests/fixtures/extensions/api-extension/PhylumExt.toml:2]: Yeah, that was the goal, initially. Don't think we need to split this test over many extensions, we can have just one that covers all the available API -- but again, we need a mock to test this against.

File: cli/tests/fixtures/extensions/api-extension/main.ts
Comment (andreaphylum) [reply cli/tests/fixtures/extensions/api-extension/main.ts:1]: Indeed -- I think we have a couple options wrt the shim: either we find a way of injecting it statically, or we distribute it and provide documentation. In both cases we'd have to at least provide some type-checking tests for the shim. Ultimately I'm not confident about where the ideal spot for the shim file would be.

File: cli/tests/fixtures/extensions/api-extension/main.ts
Comment (andreaphylum) [reply cli/tests/fixtures/extensions/api-extension/main.ts:5]: I'm not sure there is any part of the Deno API that requires testing from our side; what we do need to test is whether the usable ops are well-formed and well-typed, and `PhylumApi` is a fundamental part of it. We could achieve unit tests by constructing Deno runtimes and running extensions manually while an instance of the mock server is running in-memory for the duration of the tests, and env variables point to that. This probably requires some dedicated support code (e.g. a thread with the mock server running alongside tests and exiting once they're over).

File: cli/tests/fixtures/extensions/api-extension/phylum-api.ts
Comment (andreaphylum) [reply cli/tests/fixtures/extensions/api-extension/phylum-api.ts:8]: I agree that would be great value added; I think we can get to a good design for that organically, as we develop the extensions.

File: cli/src/deno.rs
Comment (andreaphylum) [reply cli/src/deno.rs]: You're right -- I think that's a remnant of an attempt I made which required me to hold the `RefCell` guard for longer. Corrected this.

File: cli/tests/fixtures/extensions/api-extension/phylum-api.ts
Comment (cd-work) [reply cli/tests/fixtures/extensions/api-extension/phylum-api.ts:4]: > in the end a valid URL was required every time

You need a URI for modules to be imported, that is correct. But that URI doesn't actually need to be a valid file path. I tried that out when I was messing around with the WASM stuff. The path itself is resolved and read by the module loader, which is provided by us, so we should be able to inject from there.

File: cli/tests/extensions.rs
Comment (cd-work) [reply cli/tests/extensions.rs:290]: I mean other tests already do the output check. I think that doesn't require waiting.

If you're holding off on fleshing out this test because you cannot implement it yet, then I wouldn't add the test at all rather than adding a useless test that doesn't do what it's supposed to yet.

File: cli/tests/fixtures/extensions/api-extension/phylum-api.ts
Comment (andreaphylum) [reply cli/tests/fixtures/extensions/api-extension/phylum-api.ts:4]: Oh, I had not understood that. So I guess we can make a special case in the module loader which returns our shim code if the module is e.g. `phylum-api`. I'll try that.

File: cli/tests/extensions.rs
Comment (andreaphylum) [reply cli/tests/extensions.rs:290]: Indeed, it is more of a placeholder atm, though I did use the output for manually testing some concepts. We can remove it.

File: cli/tests/fixtures/extensions/api-extension/phylum-api.ts
Comment (cd-work) [reply cli/tests/fixtures/extensions/api-extension/phylum-api.ts:4]: > Oh, I had not understood that. So I guess we can make a special case in the module loader which returns our shim code if the module is e.g. phylum-api. I'll try that.

Yeah that's actually exactly what I was thinking, but I haven't tried it. All I tried was passing a fake module URI to the module loader when loading just a single module with the default loader just directly from a string. So it should be possible to have a module loader that accepts non-paths.

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs:17]: I usually always order imports as std, crate, foreign as I feel it makes for a more intuitive hierarchy. Unless this is a common idiom I wasn't aware of, I'd defer this to a style guide. Is such a rule supported by linters? Would be nice to have.

Comment (cd-work) [reply cli/src/commands/extensions/api.rs:17]: Rustfmt can automatically sort and what I suggested is the default when enabling that unstable rustfmt option. It's also what is used by every project I'm aware of.

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs:64]: It might not be clear from this, but all ops do use `ExtensionState` as it is the object through which `PhylumApi` is accessed, so we need to provide this in `OpState` at runtime initialization. This specific function is called by most if not all ops, so not having it would create a lot of noise. We can look for alternative forms (`impl From<&mut OpState>` / `impl TryFrom<&mut OpState>` + `impl Deref<Target = PhylumApi>`?), but the substance of it is necessary.

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs:90]: I think there is value in documenting this behavior somewhere, as `RefCell`s and pinning aren't mechanisms everyone is familiar with. Probably moving this above the functions and expanding on it would be better.

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs:64]: Actually, we may not use `impl From<T>` in this case as the associated function is asynchronous (that's probably the reason I used this form in the first place).

File: cli/tests/fixtures/extensions/api-extension/main.ts
Comment (andreaphylum) [reply cli/tests/fixtures/extensions/api-extension/main.ts:5]: Yeah, I think so -- I created #450 to track this.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [reply cli/src/commands/extensions/api.rs:98]: That's a solid lint.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [reply cli/src/commands/extensions/api.rs:63]: The problem is just that `ExtensionState` here is neither input nor output, it's just a transitory maybe thing that should probably exist on `OpState`. There's various other ways to do this, either using traits or putting it on other structs, providing an unassociated method would work too. But putting it on `ExtensionState` doesn't feel right.

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs]: I did move the comment outside of that API and above all of the `#[op]` annotated methods, as a general development guide (I purposefully avoided making that a doc comment as it's solely intended for people who will write new ops). `RefCell` and `Pin` are mechanisms that are tightly bound to Rust itself and require a nontrivial amount of time to figure out (or at least, they did for me), so I feel that a few lines explaining the design choice are warranted there, so that future maintainers won't need to spend time figuring those out. Even better would be wrapping everything in a "smart pointer" of sorts that will be able to control the lifetime of the held reference.

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs:63]: I would very much like using known traits like `TryFrom` and `Deref`, but this is not possible as the crux of the issue is the await point in the middle of the chain. An unassociated function is really the only other viable option here but I don't really agree that that would provide semantic advantages over the current solution. I can try building a "smart pointer"/guard object to handle pinning like I mentioned in the other comment, which would probably be a better solution overall.

File: cli/src/commands/extensions/mod.rs
Comment (cd-work) [inline cli/src/commands/extensions/mod.rs:5]: Honestly not a big fan of `pub use` in general. I feel like things are going to be a ton more obvious if we just make the modules public?

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [inline cli/src/commands/extensions/api.rs:25]: Does this need to be `pub(crate)`?

File: cli/src/deno.rs
Comment (cd-work) [inline cli/src/deno.rs:23]: I don't think `deps` is a great variable name here? It says basically nothing about what it is or how it is used. I'd either name it something like shared state, or phylum api.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [inline cli/src/commands/extensions/api.rs:53]: Not a fan of the `t` variable name here again. I'd much rather call it `api` or `api_future`.

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [inline cli/src/commands/extensions/api.rs:60]:
```suggestion
/// Wraps a shared, counted reference to the `PhylumApi` object.
///
/// The reference can be safely extracted from `Rc<RefCell<OpState>>` and cloned; it will not require mutable access to the owning `RefCell`, so the mutable borrow to it may be dropped.
```

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [inline cli/src/commands/extensions/api.rs:64]: Technically with a boxed future you probably could, but I agree that it's not worth it.

**Review by cd-work — CHANGES_REQUESTED** [review]

File: cli/src/commands/extensions/api.rs
Comment (cd-work) [inline cli/src/commands/extensions/api.rs:28]: When I asked if this needs to be `pub(crate)` I was asking if this needs to be `pub` at all. Based on my tests, this does not need to be public.

**Review by cd-work — CHANGES_REQUESTED** [review]

**Review by cd-work — APPROVED** [review]

**Review by cd-work — APPROVED** [review]

File: cli/src/commands/extensions/api.rs
Comment (kylewillmon) [inline cli/src/commands/extensions/api.rs:214]: Calling this `analyze_package()` makes it sound like it will submit the package for analysis, which is not true. The `PhylumApi` function is `get_package_details()` and I think that name (or a variant of it) better matches the actual functionality.

File: cli/src/lockfiles/mod.rs
Comment (kylewillmon) [inline cli/src/lockfiles/mod.rs:27]: Changing this forces all of the callers to use `Path::new()` just so that one of the callers (`extensions/api.rs`) doesn't have to call `std::fs::read_to_string()`...

I think it would have made much more sense to have the extension code read the file and then call `parse(data)` instead of `parse_file(path)`.

**Review by kylewillmon — APPROVED** [review]

File: cli/src/lockfiles/mod.rs
Comment (andreaphylum) [reply cli/src/lockfiles/mod.rs:27]: I agree that having the generics looks infinitely better; the problem is not `parse_file`, per se, but the fact that a trait object can not be instantiated in the first place if one of the trait methods have a bound on `Self: Sized`.

A trait object is by definition `!Sized` and a method `where Self: Sized` cannot exist at all, hence the trait cannot be fully implemented on a trait object. Unfortunately, we do need to construct a trait object as a consequence of moving away from closures for the parsers.

Comment (kylewillmon) [reply cli/src/lockfiles/mod.rs:27]: Right. But this function is just a convenience function around `Parse::parse()`, which has no `Sized` requirement and can be called on trait objects.

So any code that needs to use trait objects can and should just call `Parse::parse()`

Comment (andreaphylum) [reply cli/src/lockfiles/mod.rs]: Indeed, for some reason I believed you couldn't create trait objects _at all_ with sized-constrained trait methods, but I have now verified I was wrong. Reverted this change in the latest commit.

File: cli/src/commands/extensions/api.rs
Comment (andreaphylum) [reply cli/src/commands/extensions/api.rs]: Renamed this in the latest commit.

**Review by kylewillmon — APPROVED** [review]

**Review by cd-work — APPROVED** [review]

**Pull request merged**
